pkg/util/spill/agg_spill_util.go

```diff
@@ -0,0 +1,333 @@
+}
+
+// SerializeInterface serialize interface type and return the number of bytes serialized
+func SerializeInterface(value interface{}, varBuf *[]byte, tmpBuf []byte) {
+	switch v := value.(type) {
+	case bool:
+		*varBuf = append(*varBuf, BoolType)
+		if v {
+			*varBuf = append(*varBuf, byte(1))
+		} else {
+			*varBuf = append(*varBuf, byte(0))
+		}
+	case int64:
+		*varBuf = append(*varBuf, Int64Type)
+		SerializeInt64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+	case uint64:
+		*varBuf = append(*varBuf, Uint64Type)
+		SerializeUint64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:uint64Len]...)
+	case float64:
+		*varBuf = append(*varBuf, FloatType)
+		SerializeFloat64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:float64Len]...)
+	case string:
+		*varBuf = append(*varBuf, StringType)
+		vLen := int64(len(v))
+		SerializeInt64(vLen, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+		*varBuf = append(*varBuf, v...)
+	case types.BinaryJSON:
+		*varBuf = append(*varBuf, BinaryJSONType)
+		varBufLenBeforeSerializeJSON := int64(len(*varBuf))
+
+		// Add padding for seialization
+		*varBuf = append(*varBuf, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
+		*varBuf = SerializeBinaryJSON(&v, *varBuf, varBufLenBeforeSerializeJSON)
+
+	case types.Opaque:
+		*varBuf = append(*varBuf, OpaqueType)
+		*varBuf = append(*varBuf, v.TypeCode)
+		SerializeInt64(int64(len(v.Buf)), tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+		*varBuf = append(*varBuf, v.Buf...)
+	case types.Time:
+		*varBuf = append(*varBuf, TimeType)
+		SerializeUint64(uint64(v.CoreTime()), tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:uint64Len]...)
+		SerializeUint8(v.Type(), tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:uint8Len]...)
+		SerializeInt(v.Fsp(), tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:intLen]...)
+	case types.Duration:
+		*varBuf = append(*varBuf, DurationType)
+		SerializeInt64(int64(v.Duration), tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+		SerializeInt(v.Fsp, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:intLen]...)
+	default:
+		panic("Agg spill encounters an unexpected interface type!")
+	}
+}
+
+// SerializeBinaryJSON serializes Set type
+func SerializeBinaryJSON(json *types.BinaryJSON, varBuf []byte, startPos int64) []byte {
+	varBuf[startPos] = json.TypeCode
+	valueLen := len(json.Value)
+	SerializeInt(valueLen, varBuf[startPos+byteLen:])
+	varBuf = varBuf[:startPos+byteLen+intLen]
+	return append(varBuf, json.Value...)
+}
+
+// SerializeSet serializes Set type
+func SerializeSet(value *types.Set, varBuf []byte, startPos int64) []byte {
+	SerializeUint64(value.Value, varBuf[startPos:])
+	varBuf = varBuf[:startPos+uint64Len]
+	return append(varBuf, value.Name...)
+}
+
+// SerializeEnum serializes Set type
+func SerializeEnum(value *types.Enum, varBuf []byte, startPos int64) []byte {
+	SerializeUint64(value.Value, varBuf[startPos:])
+	varBuf = varBuf[:startPos+uint64Len]
+	return append(varBuf, value.Name...)
+}
```

Comment:
> The interface definition of SerializeJSON/Set/Enum is different from the definition of SerializeFloat32/X. It's not elegant and hard to use.

I think we can pass pointer of varBuf, so that it's needless to return data.